Question: Implement Multiple Linear Regression for the attached dataset by splitting the dataset into training and test dataset..

1.  Apply MLR using scikit learn library function and predict the accuracy.

2. Apply MLR without using any library functions and predict the accuracy

3. Compare the accuracy of 1. and 2 and justify the result.

#Using sklearn

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error,mean_absolute_error
from sklearn import preprocessing
from sklearn.metrics import r2_score

In [ ]:
df1=pd.read_csv('/content/Admission_Predict.csv')
X = df1.drop(['Serial No.','Chance of Admit '],axis= 1)
y = df1['Chance of Admit ']
print(X)
print(y)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)
model = LinearRegression()
model.fit(X_train,y_train)
predictions = model.predict(X_test)
print('mean_squared_error : ', mean_squared_error(y_test, predictions))
print('mean_absolute_error : ', mean_absolute_error(y_test, predictions))
print("R2-score: %.2f (RMSE):" , r2_score(y_test, predictions) )

     GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  Research
0          337          118                  4  4.5   4.5  9.65         1
1          324          107                  4  4.0   4.5  8.87         1
2          316          104                  3  3.0   3.5  8.00         1
3          322          110                  3  3.5   2.5  8.67         1
4          314          103                  2  2.0   3.0  8.21         0
..         ...          ...                ...  ...   ...   ...       ...
395        324          110                  3  3.5   3.5  9.04         1
396        325          107                  3  3.0   3.5  9.11         1
397        330          116                  4  5.0   4.5  9.45         1
398        312          103                  3  3.5   4.0  8.78         0
399        333          117                  4  5.0   4.0  9.66         1

[400 rows x 7 columns]
0      0.92
1      0.76
2      0.72
3      0.80
4      0.65
       ... 
395    0.82
396 

#Without Using sklearn

In [ ]:
x = df1.iloc[:,:-1].values
y = df1.iloc[:,-1].values
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x = sc.fit_transform(x)
Xtrain, Xtest, Ytrain, Ytest = train_test_split( x, y, test_size = 0.3,random_state=101)

In [ ]:
def ChanceFunction(x,y,w,b):
    chance = np.sum((((x.dot(w) + b) - y) ** 2) / (2*len(y)))
    return chance
def GradientDescent(x, y, w, b, learning_rate, epochs):
    cost_list = [0] * epochs
    for epoch in range(epochs):
        z = x.dot(w) + b
        loss = z - y   
        weight_gradient = x.T.dot(loss) / len(y)
        bias_gradient = np.sum(loss) / len(y)   
        w = w - learning_rate*weight_gradient
        b = b - learning_rate*bias_gradient
        cost = ChanceFunction(x, y, w, b)
        cost_list[epoch] = cost      
        if (epoch%(epochs/10)==0):
            print("Chance is : ",cost)      
    return w, b, cost_list
w, b, c= GradientDescent(Xtrain, Ytrain, np.zeros(Xtrain.shape[1]), 0, 0.002,epochs=15000)

Chance is :  0.281117283352708
Chance is :  0.002508346736662972
Chance is :  0.0017411337540491957
Chance is :  0.0017298518526890575
Chance is :  0.0017281469827862628
Chance is :  0.0017276021382782412
Chance is :  0.0017273960195330464
Chance is :  0.0017273139715735584
Chance is :  0.0017272803975796301
Chance is :  0.0017272663944848153


In [ ]:
def predict(X, w, b):
    return X.dot(w) + b
y_pred = predict(Xtest, w, b)
def r2score(y_pred, y):
    rss = np.sum((y_pred - y) ** 2)
    tss = np.sum((y-y.mean()) ** 2)
    
    r2 = 1 - (rss / tss)
    return r2
r2score(y_pred, Ytest)

0.761289750994438

#Accurcy with using sklearn is 73.9% and Accuracy without sklearn is 76.1%